src/linear_algebra/orthogonal_group.lean

```diff
@@ -0,0 +1,159 @@
+/-
+Copyright (c) 2021 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Shing Tak Lam
+-/
+import linear_algebra.matrix
+import linear_algebra.nonsingular_inverse
+import data.real.basic
+
+/-!
+# The Orthogonal Group O(n)
+
+This file defines elements of the orthogonal group `orthogonal_group n`, also written `O(n)`,
+consisting of all `n` by `n` real matrices where the transpose is the inverse. In addition,
+we define the group structure on `orthogonal_group n` and the embedding into the general linear
+group `general_linear_group ℝ (n → ℝ)`.
+
+## Main Definitions
+
+ * `matrix.orthogonal_group` is the type of matrices where the transpose is the inverse
+ * `matrix.orthogonal_group.group` is the group structure (under multiplication)
+ * `matrix.orthogonal_group.embedding_GL` is the embedding `O(n) → GLₙ(ℝ)`
+
+## References
+
+ * https://en.wikipedia.org/wiki/Orthogonal_group
+
+## Tags
+
+matrix group, group, orthogonal group
+
+-/
+
+namespace matrix
+
+universe u
+open_locale matrix
+open linear_map
+
+section
+
+variables (n : Type u) [fintype n] [decidable_eq n]
+
+/-- `orthogonal_group n` is the group of `n` by `n` matrices where the transpose is the inverse. -/
+def orthogonal_group := {M : matrix n n ℝ // matrix.mul Mᵀ M = 1}
```

Comment:
Does it make sense to define this as a `subgroup`, which provides the `group` and `ext_*` structures automatically? The downside of that is that dot notation doesn't work, but you could compromise on:
```suggestion
/-- `orthogonal_subgroup n` is the subgroup of `n` by `n` matrices where the transpose is the inverse. -/
def orthogonal_subgroup : subgroup (matrix n n ℝ) :=
{ carrier := {M : matrix n n ℝ // matrix.mul Mᵀ M = 1}, mul_mem' := sorry, one_mem' := sorry, inv_mem' := sorry}

/-- `orthogonal_group n` is a convenience alias for the coe_sort of `orthogonal_subgroup`, to enable dot notation. -/
def orthogonal_group : Type* := orthogonal_subgroup n
```

Reply:
Umm, there is no group instance on `matrix n n \R`? Unless if you mean `GL_n(\R)`, but that's defined based on linear maps and not matrices.

Reply:
Whoops, that was dumb of me. There's (or could be) a `monoid` instance though, right? So you could create a `submonoid`, and then manually endow it with the inverse structure.